### PR TITLE
Tests: Add tests for utils.py

### PR DESCRIPTION
**Pull request by SchoolGuy**
Since `utils.py` is kind of an import part of our project I wanted to explicitly test everything we have in there to keep it stable and working. Since this is kind of a mad file regarding the length it will take a while until it is ready.

### CODE REVIEW

File: tests/utils_test.py [hunk 1]
```diff
@@ -0,0 +1,1087 @@
+import os
+
+import pytest
+from netaddr.ip import IPAddress
+
+from cobbler import utils
+from cobbler.cexceptions import CX
+
+
+def test_pretty_hey():
```
Comment (watologo1) [inline tests/utils_test.py:10]: test_pretty_hex?

**Review by watologo1 — CHANGES_REQUESTED** [review]
Are all these tiny unusued utils tests needed?
Why and where?

**Comment by watologo1** [comment]
It would also be better to split this in fixes and and the new tests (file).
Beside the one "r" that slipped in, these fixes can/should just go in.

**Comment by SchoolGuy** [comment]
@watologo1 The didn't slip in. It was put there intentional since this is now a raw docstring which enables Python to handle the docstring differently. The fixes and the tests are dependent. Since I am not rewriting a whole module I would like to keep this together. Also this is not finished and this the empty calls.

**Comment by hbokh** [comment]
@SchoolGuy Any idea what is happening here?
https://travis-ci.com/github/cobbler/cobbler/jobs/395116973#L1929-L1932

**Comment by SchoolGuy** [comment]
@hbokh Nope but I will have a look locally and push a fix later.

File: tests/utils_test.py (hunk 1 above)
Comment (SchoolGuy) [reply tests/utils_test.py:10]: Will be fixed with the next commit.

File: cobbler/utils.py [hunk 2]
```diff
@@ -1097,17 +1093,19 @@ def is_safe_to_hardlink(src, dst, api):
 
 
 def hashfile(fn, lcache=None, logger=None):
-    """
+    r"""
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply cobbler/utils.py:1096]: That is intentional as this is then a raw docstring.

**Comment by SchoolGuy** [comment]
@watologo1 @hbokh @dansou901 @Conan-Kudo @loreson I would love if we could review this rather sooner then later. I will clean up the git history in the next few minutes so the CI will run again.

**Comment by dansou901** [comment]
@SchoolGuy I won't be of much help reviewing this, as I don't speak python. I'm leaving this one to the others.

**Comment by SchoolGuy** [comment]
@nodeg & @m-rey your opinions are also valid in this matter I think, although you currently don't belong to the project officially.

**Comment by opoplawski** [comment]
I don't have much to add other that I'm glad that folks are writing tests.

**Review by nodeg — CHANGES_REQUESTED** [review]
I left some suggestions.

File: cobbler/utils.py [hunk 3]
```diff
@@ -879,14 +874,14 @@ def dict_to_string(_dict):
...
     Sync files from src to dst. The extra arguments specified by args are appended to the command.
 
     :param src: The source for the copy process.
     :param dst: The destination for the copy process.
-    :param args: The extra arguments are appended to our standard arguements.
+    :param args: The extra arguments are appended to our standard arguments.
     :param logger: The logger to audit the action with.
-    :param quiet: If "True" no progress is reported. If "False" then progress will be reported by rsync.
+    :param quiet: If "True" no progress is reported. If ``False`` then progress will be reported by rsync.
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply cobbler/utils.py:884]: Will be fixed in the next commit.

File: cobbler/utils.py [hunk 4]
```diff
@@ -2145,10 +2141,10 @@ def to_string_from_fields(item_dict, fields, interface_fields=None):
     """
     item_dict is a dictionary, fields is something like item_distro.FIELDS
 
-    :param item_dict: Not known what this is needed for exactly.
-    :param fields: Not known what this is needed for exactly.
-    :param interface_fields: Not known what this is needed for exactly.
-    :return: Not known what this is returning exactly.
+    :param item_dict: The dictionary representation of a Cobbler Item.
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply cobbler/utils.py:2144]: Will be fixed in the next commit.

File: cobbler/utils.py [hunk 5]
```diff
@@ -2378,6 +2371,7 @@ def strip_none(data, omit_none=False):
 
     return data
 
+
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply cobbler/utils.py:2374]: I removed the comment line. It makes no sense indeed.

File: cobbler/utils.py [hunk 6]
```diff
@@ -2405,20 +2399,20 @@ def revert_strip_none(data):
...
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply cobbler/utils.py:2402]: This makes no sense indeed.

File: tests/utils_test.py [hunk 7]
```diff
@@ -0,0 +1,1240 @@
...
+    # Arrange
+    testfolder = "/dev/shm/testfoldercreation"
+    testmode = 0o600
+
+    try:
+        shutil.rmtree(testfolder)
+    except OSError:
+        pass
+
+    # Pre Assert to check that this actually does something
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply tests/utils_test.py:599]: Will be fixed in the next commit.

File: tests/utils_test.py [hunk 8]
```diff
@@ -0,0 +1,1240 @@
...
+    cache_src = ""
+    cache_dst = ""
+    api = None
+
+    # Act
+    utils.cachefile(cache_src, cache_dst, api=api)
+
+    # Assert
+    # TODO: Check .link_cache folder exists and the link cache file in it
+    # TODO: Assert file exists on the cache destination
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply tests/utils_test.py:487]: Will be fixed in the next commit.

File: tests/utils_test.py [hunk 9]
```diff
@@ -0,0 +1,1240 @@
...
+def test_run_triggers():
+    # Arrange
+    api = CobblerAPI()
+    globber = ""
+
+    # Act
+    utils.run_triggers(api, None, globber)
+
+    # Assert
+    # TODO: How the heck to we check this actually did what it is supposed to do?
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply tests/utils_test.py:418]: Will be fixed in the next commit.

File: tests/utils_test.py [hunk 10]
```diff
@@ -0,0 +1,1240 @@
...
+    time_before_write = time.time()
+
+    # Act
+    result = utils.update_settings_file(settings_data)
+
+    # Assert
+    assert result
+    # This should work the following: The time of modifying the settings file should be greater then the time taken
+    # before modifying it. Thus the value of the subtraction of both should be great 0. If writing to the files does not
+    # work this is smaller then 0. The content is a yaml file thus we don't want to test if writing a yaml file is
```
[the remark this answers is not in the document]
Comment (SchoolGuy) [reply tests/utils_test.py:286]: Will be fixed in the next commit.

**Comment by SchoolGuy** [comment]
Codacy may fail since the subprocess calls are intentional and we havn't had the time to configure the other linters regarding the Doc Comment styles.

**Review by watologo1 — APPROVED** [review]
IMO changing comments to raw strings to show code elements:
""" dict() is a method

to
r""" ``dict()`` is a method

is IMO somewhat overdosed. Just keep the rest. There is more important things to do...
The submitrequest looks fine and correct to me. If, then maybe too correct ;)

**Review by nodeg — APPROVED** [review]
LGTM.

**Comment by SchoolGuy** [comment]
Since we two approvals. I will merge now.

**Pull request merged**